### PR TITLE
HIP-584 Historical: Fix historical NFT/Fungible token balance

### PR DESCRIPTION
**Pull request by 0xivanov**
**Description**:

This PR includes the following fixes:
- get historical fungible token balance from `TokenBalanceRepository`
- get historical NFT balance from `NftRepository`. Historical NFT balance should be obtained from `account.getOwnedNfts()`, as explained in the code comments
- fix `findHistoricalTokenBalanceUpToTimestamp` in edge case where `balance_snapshot` and/or `base` queries do not match anything in the db. Currently reproducible in acceptance tests:
1. create account
2. make fungible token transfer +50
3. make fungible token transfer +50
4. make historical call with timestamp between the token transfers, expected balance is 50, but returns 0

I have observed that the state in db after these operations is the following:
`account_balance` - no entry for this accountId
`token_balance` - no entry for this accountId
`token_transfers` - 2 entries for this accountId, one for first transfer (50) and one for the second transfer (50)

`balance_snapshot` and `base` match nothing because there is no entry for this accountId before the given timestamp.
This causes the statement `tt.consensus_timestamp > s.consensus_timestamp` to be false and cannot match any token transfers, always returning 0 due to the coalesce at the end of the query.
NB. `token_balance` and `account_balance` entries get persisted in ~8 minute intervals so it is not certain that there is entry in the db after the first transfer.

Current solution:
If `balance_snapshot` and/or `base` do not match anything, find token transfers that occured after the account creation and before the given timestamp:
```sql
tt.consensus_timestamp >= coalesce((select consensus_timestamp from base), accountCreatedTimestamp)
```
Changed `>` to `>=` ensures the inclusion of the token transfer that occurred when the account was created.

**Optimisation proposals:**
- remove the `balance_snapshot` part of the query as it is not relevant to the token balances

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java [hunk 1]
```diff
@@ -62,6 +65,40 @@ public class TokenRelationshipDatabaseAccessor extends DatabaseAccessor<Object,
...
+     *
+     * Non-historical Call:
+     * The same principle as NFT applies here, and the balance is obtained from `account.getOwnedNfts()`.
+     *
+     * Historical Call:
+     * In historical block queries, since the `token_account` table lacks historical state for fungible balances,
+     * the balance is determined from the `token_balance` table using the `findHistoricalTokenBalanceUpToTimestamp` query.
+     */
+    private Long getBalance(Account account, Token token, TokenAccount tokenAccount, final Optional<Long> timestamp) {
+        if (token.getType().equals(TokenType.NON_FUNGIBLE_UNIQUE)) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java:92]: I believe this method is only applicable for fungible token amount and we shouldn't have any logic for nfts.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java [hunk 2]
```diff
@@ -57,37 +57,83 @@ public interface TokenBalanceRepository extends CrudRepository<TokenBalance, Tok
...
+                        )
+                        select coalesce((select balance from base), 0) + coalesce((select amount from change), 0)
                     """,
             nativeQuery = true)
-    Optional<Long> findHistoricalTokenBalanceUpToTimestamp(long tokenId, long accountId, long blockTimestamp);
+    Optional<Long> findHistoricalTokenBalanceUpToTimestamp(
+            long tokenId, long accountId, long blockTimestamp, long accountCreatedTimestamp);
+    // version 1 - removed account_balance part
+
+    //                    with base as (
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java:93]: Probably this commented code from here to the bottom is obsolete?

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java (hunk 1 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java:92]: In `RedirectViewExecutor` we call `tokenAccessor.balanceOf(...)` when we have `ABI_ID_ERC_BALANCE_OF_TOKEN` signature which gets the balance for the token-account from `TokenRelationshipDatabaseAccessor`, regardless of the token type.
Check out
` Then I verify that historical data for "FUNGIBLE_HISTORICAL" is returned via balanceOf`
and
`Then I verify that historical data for "NFT_HISTORICAL" is returned via balanceOf`
from https://github.com/hashgraph/hedera-mirror-node/pull/7418/files
In the acceptance tests we get the historical NFT balance from this method, that is why i changed it

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java (hunk 2 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java:93]: Wrote the 2 queries there for easier review process. Both versions apply the fixes that i mention in the description, but version 1 is my proposal to remove `account_balance` from the query.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java (hunk 1 above)
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java:92]: I see, then it seems OK to return the `owned nfts` from the `owner`, in case of NFT.

**Review by IvanKavaldzhiev — DISMISSED** [review]
LGTM

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java [hunk 3]
```diff
@@ -62,6 +65,40 @@ public class TokenRelationshipDatabaseAccessor extends DatabaseAccessor<Object,
...
+     * NFT Balance Explanation:
+     *
+     * Non-historical Call:
+     * When querying the latest block, the NFT balance from `account.getOwnedNfts()`
+     * matches the balance from `tokenAccount.getBalance()`.
+     * Therefore, the NFT balance is obtained from `account.getOwnedNfts()`.
+     *
+     * Historical Call:
+     * In historical block queries, as the `token_account` and `token_balance` tables lack historical state for NFT balances,
+     * the NFT balance is retrieved from `account.getOwnedNfts()`, previously set in `AccountDatabaseAccessor.getOwnedNfts()`.
```
Comment (xin-hedera) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java:80]: this seems to imply that `account` and `timestamp` are correlated?

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java [hunk 4]
```diff
@@ -62,6 +65,40 @@ public class TokenRelationshipDatabaseAccessor extends DatabaseAccessor<Object,
...
+     *
+     * Fungible Token Balance Explanation:
+     *
+     * Non-historical Call:
+     * The same principle as NFT applies here, and the balance is obtained from `account.getOwnedNfts()`.
```
Comment (xin-hedera) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java:85]: think it should be `tokenAccount.getBalance()` instead of `account.getOwnedNfts()`

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java [hunk 5]
```diff
@@ -57,37 +57,83 @@ public interface TokenBalanceRepository extends CrudRepository<TokenBalance, Tok
...
-                        select sum(amount) as amount
-                        from token_transfer as tt, balance_snapshot as s
-                        where
-                          token_id = ?1 and
-                          account_id = ?2 and
-                          tt.consensus_timestamp > s.consensus_timestamp and
-                          tt.consensus_timestamp <= ?3
-                    )
-                    select coalesce((select balance from base), 0) + coalesce((select amount from change), 0)
+                        with base as (
```
Comment (xin-hedera) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java:68]: please use the same CTE in the historical crypto balance PR to find the latest balance snapshot timestamp <= blockTimestamp and use the balance snapshot timestamp to get the base token balance

I don't see the value of adding `accountCreatedTimestamp` parameter.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java [hunk 6]
```diff
@@ -57,37 +57,83 @@ public interface TokenBalanceRepository extends CrudRepository<TokenBalance, Tok
...
+                                consensus_timestamp <= ?3
+                            order by consensus_timestamp desc
+                            limit 1
+                        ), change as (
+                            select sum(amount) as amount
+                            from token_transfer as tt
+                            where
+                                token_id = ?1 and
+                                account_id = ?2 and
+                                tt.consensus_timestamp >= coalesce((select consensus_timestamp from base), ?4) and
```
Comment (xin-hedera) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java:83]:
```suggestion
                                tt.consensus_timestamp > coalesce((select consensus_timestamp from base), ?4) and
```

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java (hunk 3 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java:80]: Not really, instead it implies that this specific `timestamp` was also passed when loading this account in the cache from the db and we got it's historical nft balance from there (with the same timestamp passed here)

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java (hunk 4 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java:85]: Correct

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java (hunk 6 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java]: fixed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java (hunk 5 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java]: fixed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java (hunk 4 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java]: Separated the logic for better readability

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java [hunk 7]
```diff
@@ -62,6 +66,55 @@ public class TokenRelationshipDatabaseAccessor extends DatabaseAccessor<Object,
...
+     * Non-historical Call:
+     * When querying the latest block, the NFT balance from `account.getOwnedNfts()`
+     * matches the balance from `tokenAccount.getBalance()`.
+     * Therefore, the NFT balance is obtained from `account.getOwnedNfts()`.
+     * Historical Call:
+     * In historical block queries, as the `token_account` and `token_balance` tables lack historical state for NFT balances,
+     * the NFT balance is retrieved from `account.getOwnedNfts()`, previously set in `AccountDatabaseAccessor.getOwnedNfts()`.
+     */
+    private Long getNftBalance(final Account account) {
+        return account.getOwnedNfts();
```
Comment (0xivanov) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java:91]: After further testing, discovered that this method returns the number of all owned nfts and not for this specific nft id. Should change the implementation

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java (hunk 3 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java]: This logic is now changed, please have a look.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java (hunk 7 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java]: fixed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java [hunk 8]
```diff
@@ -110,7 +110,7 @@ select count(*)
                             select token_id
                             from nft
                             where account_id = :accountId
-                                and lower(timestamp_range) <= :blockTimestamp
+                                and timestamp_range @> :blockTimestamp
```
Comment (0xivanov) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java:113]: @xin-hedera Can you please explain where this operator `@>` should be used?
Currently we are using combination of `@>` and `lower(timestamp_range) <=` , which one is correct?
From what I have tested in acceptance tests - using `@>` has correct behaviour, in contrast - `lower(timestamp_range) <=` returns incorrect values occasionally.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java [hunk 9]
```diff
@@ -64,29 +66,33 @@ public interface TokenBalanceRepository extends CrudRepository<TokenBalance, Tok
...
-                    select coalesce((select balance from base), 0) + coalesce((select amount from change), 0)
+                        with balance_timestamp as (
+                        select consensus_timestamp
+                        from account_balance
+                        where account_id = 2 and
+                            consensus_timestamp > ?3 - 2678400000000000 and consensus_timestamp <= ?3
+                        order by consensus_timestamp desc
+                        limit 1
+                        ), base as (
+                            select tb.balance, tb.consensus_timestamp
```
Comment (xin-hedera) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java:77]:
```suggestion
                            select tb.balance
```
Using the token's balance timestamp when token balance info is deduped can increase the timestamp range when querying token_transfer table and it's completely unnecessary.

For example, based on account 2, the balance timestamp is 5000. In the `base` CTE, the latest deduped token balance info at or before timestamp 5000 has `balance = 20, consensus_timestamp = 1000`.

This indicates for all the balance snapshot timestamps in the range `[1000,  5000]`, the token's. balance is always 20. To reconcile and find the token balance change, we should query token_balance table in the timestamp range `(5000,  blockTimestamp]`,  instead of `(1000, blockTimestamp]`, because the deduped balance info implies in the timestamp range `(1000, 5000]`, there is no token transfer for the account and token at all.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java [hunk 10]
```diff
@@ -64,29 +66,33 @@ public interface TokenBalanceRepository extends CrudRepository<TokenBalance, Tok
...
+                                tb.consensus_timestamp <= bt.consensus_timestamp
+                            order by consensus_timestamp desc
+                            limit 1
+                        ), change as (
+                            select sum(amount) as amount
+                            from token_transfer as tt
+                            where
+                                token_id = ?1 and
+                                account_id = ?2 and
+                                tt.consensus_timestamp > coalesce((select consensus_timestamp from base), 0) and
```
Comment (xin-hedera) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java:92]:
```suggestion
                                tt.consensus_timestamp > coalesce((select consensus_timestamp from balance_timestamp), 0) and
```

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java (hunk 8 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java:113]: Does this mean we can use `timestamp_range @>` everywhere?

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java (hunk 9 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java]: Fixed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java (hunk 10 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java]: Fixed

**Comment by 0xivanov** [comment]
Some integration tests need to be adjusted.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java (hunk 8 above)
Comment (xin-hedera) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/NftRepository.java:113]: yes.

In some places, we see the pattern in the query for the history table `where $1 >= lower(timestamp_range) and ... order by lower(timestamp_range) limit 1`, that's a performance optimization because the query can utilize the index, for example, for `token` table, the index on ` (token_id, lower(timestamp_range))`

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java [hunk 11]
```diff
@@ -62,6 +68,64 @@ public class TokenRelationshipDatabaseAccessor extends DatabaseAccessor<Object,
                 .formatted(TokenRelationship.class.getTypeName(), key.getClass().getTypeName()));
     }
 
+    /**
+     * Determines fungible or NFT balance based on block context.
+     */
+    private Long getBalance(
+            final Account account, final Token token, final TokenAccount tokenAccount, final Optional<Long> timestamp) {
+        if (token.getType().equals(TokenType.NON_FUNGIBLE_UNIQUE)) {
+            return getNftBalance(tokenAccount, timestamp, account.getCreatedTimestamp());
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/accessor/TokenRelationshipDatabaseAccessor.java:77]: thanks!

**Review by xin-hedera — APPROVED** [review]
LGTM

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java [hunk 12]
```diff
@@ -103,6 +105,7 @@ void pureCallWithCustomBlock(BlockType blockType) {
         }
     }
 
+    @Disabled
```
Comment (0xivanov) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java:108]: These 2 should get fixed here https://github.com/hashgraph/hedera-mirror-node/pull/7434

Comment (steven-sheehy) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java:108]: Now that that PR is merged, these should be removed, right?

**Review by steven-sheehy — APPROVED** [review]
LGTM besides the question about `@Disabled`

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java (hunk 12 above)
Comment (0xivanov) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceTest.java]: Removed

**Review by steven-sheehy — APPROVED** [review]
LGTM
